File: androidApp/src/test/java/com/commuto/interfacemobile/android/offer/validation/NewSwapDataValidatorTests.kt
package com.commuto.interfacemobile.android.offer.validation

import androidx.compose.runtime.mutableStateListOf
import com.commuto.interfacemobile.android.offer.Offer
import com.commuto.interfacemobile.android.offer.OfferDirection
import com.commuto.interfacemobile.android.offer.OfferState
import com.commuto.interfacemobile.android.settlement.SettlementMethod
import com.commuto.interfacemobile.android.settlement.privatedata.PrivateSWIFTData
import com.commuto.interfacemobile.android.ui.StablecoinInformationRepository
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Test
import java.math.BigDecimal
import java.math.BigInteger
import java.util.*

class NewSwapDataValidatorTests {

    /**
     * Ensures that [validateNewSwapData] validates swap data properly.
     */
    @Test
    fun testValidateNewSwapData() {
        val makerSettlementMethod = SettlementMethod(
            currency = "USD",
            price = "1.00",
            method = "SWIFT",
            privateData = Json.encodeToString(PrivateSWIFTData(
                accountHolder = "account_holder",
                bic = "bic",
                accountNumber = "account_number"
            ))
        )
        val takerSettlementMethod = SettlementMethod(
            currency = "USD",
            price = "",
            method = "SWIFT",
            privateData = Json.encodeToString(PrivateSWIFTData(
                accountHolder = "account_holder",
                bic = "bic",
                accountNumber = "account_number"
            ))
        )
        val offer = Offer(
            isCreated = true,
            isTaken = false,
            id = UUID.randomUUID(),
            maker = "0x0000000000000000000000000000000000000000",
            interfaceID = ByteArray(0),
            stablecoin = "0x8438Ad1C834623CfF278AB6829a248E37C2D7E3f", // BUSD on Hardhat,
            amountLowerBound = BigInteger.valueOf(10) * BigInteger.TEN.pow(18),
            amountUpperBound = BigInteger.valueOf(20) * BigInteger.TEN.pow(18),
            securityDepositAmount = BigInteger.valueOf(2) * BigInteger.TEN.pow(18),
            serviceFeeRate = BigInteger.valueOf(100),
            direction = OfferDirection.BUY,
            settlementMethods = mutableStateListOf(makerSettlementMethod),
            protocolVersion = BigInteger.valueOf(1),
            chainID = BigInteger.valueOf(31337),
            havePublicKey = true,
            isUserMaker = false,
            state = OfferState.OFFER_OPENED
        )
        val validatedSwapData = validateNewSwapData(
            offer = offer,
            takenSwapAmount = BigDecimal(15),
            selectedMakerSettlementMethod = makerSettlementMethod,
            selectedTakerSettlementMethod = takerSettlementMethod,
            stablecoinInformationRepository = StablecoinInformationRepository.hardhatStablecoinInfoRepo
        )
        val expectedValidatedSwapData = ValidatedNewSwapData(
            takenSwapAmount = BigInteger.valueOf(15) * BigInteger.TEN.pow(18),
            makerSettlementMethod = makerSettlementMethod,
            takerSettlementMethod = takerSettlementMethod
        )
        assertEquals(expectedValidatedSwapData, validatedSwapData)
    }
}

File: iosApp/iosAppTests/offer/validation/NewSwapDataValidatorTests.swift
//
//  NewSwapDataValidatorTests.swift
//  iosAppTests
//
//  Created by jimmyt on 8/12/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import BigInt
import web3swift
import XCTest

@testable import iosApp

class NewSwapDataValidatorTests: XCTestCase {
    /**
     Ensures that `validateNewSwapData` validates swap data properly.
     */
    func testValidateNewSwapData() throws {
        let makerSettlementMethod = SettlementMethod(currency: "USD", price: "1.00", method: "SWIFT", id: UUID().uuidString, privateData: String(decoding: try! JSONEncoder().encode(PrivateSWIFTData(accountHolder: "account_holder", bic: "bic", accountNumber: "account_number")), as: UTF8.self))
        let takerSettlementMethod = SettlementMethod(currency: "USD", price: "", method: "SWIFT", id: UUID().uuidString, privateData: String(decoding: try! JSONEncoder().encode(PrivateSWIFTData(accountHolder: "account_holder", bic: "bic", accountNumber: "account_number")), as: UTF8.self))
        let offer = Offer(
            isCreated: true,
            isTaken: false,
            id: UUID(),
            maker: EthereumAddress("0x0000000000000000000000000000000000000000")!,
            interfaceID: Data(),
            stablecoin: EthereumAddress("0x8438Ad1C834623CfF278AB6829a248E37C2D7E3f")!, // BUSD on Hardhat,
            amountLowerBound: 10 * BigUInt(10).power(18),
            amountUpperBound: 20 * BigUInt(10).power(18),
            securityDepositAmount: 2 * BigUInt(10).power(18),
            serviceFeeRate: BigUInt(100),
            direction: .buy,
            settlementMethods: [makerSettlementMethod],
            protocolVersion: BigUInt(1),
            chainID: BigUInt(31337),
            havePublicKey: true,
            isUserMaker: false,
            state: .offerOpened
        )
        let validatedData = try validateNewSwapData(
            offer: offer,
            takenSwapAmount: NSNumber(floatLiteral: 15).decimalValue,
            selectedMakerSettlementMethod: makerSettlementMethod,
            selectedTakerSettlementMethod: takerSettlementMethod,
            stablecoinInformationRepository: StablecoinInformationRepository.hardhatStablecoinInfoRepo
        )
        let expectedValidatedData = ValidatedNewSwapData(
            takenSwapAmount: 15 * BigUInt(10).power(18),
            makerSettlementMethod: makerSettlementMethod,
            takerSettlementMethod: takerSettlementMethod
        )
        XCTAssertEqual(expectedValidatedData.takenSwapAmount, validatedData.takenSwapAmount)
        
        XCTAssertEqual(expectedValidatedData.makerSettlementMethod.currency, validatedData.makerSettlementMethod.currency)
        XCTAssertEqual(expectedValidatedData.makerSettlementMethod.price, validatedData.makerSettlementMethod.price)
        XCTAssertEqual(expectedValidatedData.makerSettlementMethod.method, validatedData.makerSettlementMethod.method)
        
        XCTAssertEqual(expectedValidatedData.takerSettlementMethod.currency, validatedData.takerSettlementMethod.currency)
        XCTAssertEqual(expectedValidatedData.takerSettlementMethod.method, validatedData.takerSettlementMethod.method)
    }
}
